### DISCUSSION TITLE
"NaN loss with loss" exception when training

### DISCUSSION THREAD

**Discussion by dostee-hemin** [coqui-ai/TTS#1885] [General Q&A]
When I start the training process on my Glow TTS model, immediately on the first epoch the program fails due to a runtime error and it says "NaN loss with loss". Here's the traceback of the exception:

>Traceback (most recent call last):
  File "C:\Users\Machine Learning\Desktop\Coqui TTS\TTS\lib\site-packages\trainer\trainer.py", line 1533, in fit
    self._fit()
  File "C:\Users\Machine Learning\Desktop\Coqui TTS\TTS\lib\site-packages\trainer\trainer.py", line 1517, in _fit
    self.train_epoch()
  File "C:\Users\Machine Learning\Desktop\Coqui TTS\TTS\lib\site-packages\trainer\trainer.py", line 1282, in train_epoch
    _, _ = self.train_step(batch, batch_num_steps, cur_step, loader_start_time)
  File "C:\Users\Machine Learning\Desktop\Coqui TTS\TTS\lib\site-packages\trainer\trainer.py", line 1114, in train_step
    outputs, loss_dict_new, step_time = self._optimize(
  File "C:\Users\Machine Learning\Desktop\Coqui TTS\TTS\lib\site-packages\trainer\trainer.py", line 998, in _optimize
    outputs, loss_dict = self._model_train_step(batch, model, criterion)
  File "C:\Users\Machine Learning\Desktop\Coqui TTS\TTS\lib\site-packages\trainer\trainer.py", line 954, in _model_train_step
    return model.train_step(*input_args)
  File "C:\Users\Machine Learning\Desktop\Coqui TTS\TTS\TTS\tts\models\glow_tts.py", line 416, in train_step
    loss_dict = criterion(
  File "C:\Users\Machine Learning\Desktop\Coqui TTS\TTS\lib\site-packages\torch\nn\modules\module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "C:\Users\Machine Learning\Desktop\Coqui TTS\TTS\TTS\tts\layers\losses.py", line 529, in forward
    raise RuntimeError(f" [!] NaN loss with {key}.")
RuntimeError:  [!] NaN loss with loss.

After digging around in the code, I found out that the mean loss calculation always results in a tensor with nan values, like so:

> value of means: tensor([[[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],
        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],
        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],
        ...,
        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],
        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],
        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]]], device='cuda:0',
       grad_fn=<ToCopyBackward0>)

I don't understand where this comes from or what I can do to ignore it. Any ideas?

**Comment by dostee-hemin** [comment]
Quick update:

I found that setting `mixed_precision = False` in the config.json file removes the error, but frankly I don't know exactly what I did. 

Can someone please explain what the mixed_precision setting does and if I should use True or False? Also, if I should set it to True, how can I remove the error from arising during training?

**Reply by sudip550** [reply]
hii sir, did you found solution on it

**Comment by p0p4k** [comment]
https://docs.nvidia.com/deeplearning/performance/mixed-precision-training/index.html
